shotgun_api3/shotgun.py

```diff
@@ -2155,6 +2156,132 @@ def upload(self, entity_type, entity_id, path, field_name=None, display_name=Non
+        :param str field_name: The internal Shotgun field name on the entity to store the file in.
+            This field must be a File/Link field type.
+        :param str display_name: The display name to use for the file. Defaults to the file name.
+        :param str tag_list: comma-separated string of tags to assign to the file.
+        :param bool is_thumbnail: indicates if the attachment is a thumbnail.
+        :returns: Id of the Attachment entity that was created for the image.
+        :rtype: int
+        """
+
+        filename = os.path.basename(path)
+
+        # Step 1: get the upload url
+
+        upload_info = self._get_attachment_upload_info(is_thumbnail, filename)
+
+        # Step 2: upload the file
+
+        # Get some file info
+        fd = open(path, "rb")
+        content_type = mimetypes.guess_type(filename)[0]
+        content_type = content_type or "application/octet-stream"
+        file_size = os.fstat(fd.fileno())[stat.ST_SIZE]
+
+        # Perform the request
+        opener = urllib2.build_opener(urllib2.HTTPHandler)
+        try:
+            file = open(path, "rb")
+            request = urllib2.Request(upload_info['upload_url'], data=fd)
+            request.add_header("Content-Type", content_type)
+            request.add_header("Content-Length", file_size)
+            request.get_method = lambda: "PUT"
+            result = opener.open(request)
+        except urllib2.HTTPError, e:
+            if e.code == 500:
+                raise ShotgunError("Server encountered an internal error.\n%s\n%s\n\n" % (url, e))
+            else:
+                raise ShotgunError("Unanticipated error occurred uploading %s: %s" % (path, e))
+
+        LOG.debug("Completed upload of media to S3 of file: %s", filename)
+
+        # Step 3: create the attachment
+
+        url = urlparse.urlunparse((self.config.scheme, self.config.server,
+                                   "/upload/api_link_file", None, None, None))
+
+        params = {
+            "entity_type" : entity_type,
+            "entity_id" : entity_id,
+            "upload_link_info": upload_info['upload_info']
+        }
+
+        params.update(self._auth_params())
+
+        if is_thumbnail:
+            if field_name == "filmstrip_thumb_image" or field_name == "filmstrip_image":
+                params["filmstrip"] = True
+        else:
+            if display_name is None:
+                display_name = filename
+            # we allow linking to nothing for generic reference use cases
+            if field_name is not None:
+                params["field_name"] = field_name
+            params["display_name"] = display_name
+            # None gets converted to a string and added as a tag...
+            if tag_list:
+                params["tag_list"] = tag_list
+
+        # Create opener with extended form post support
+        opener = self._build_opener(FormPostHandler)
```

Comment:
again, this could perhaps take advantage of a shared method for running the form post style operations